light/trie.go

```diff
@@ -18,99 +18,216 @@ package light
 
 import (
 	"context"
+	"fmt"
 
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/crypto"
-	"github.com/ethereum/go-ethereum/ethdb"
 	"github.com/ethereum/go-ethereum/trie"
 )
 
-// LightTrie is an ODR-capable wrapper around trie.SecureTrie
-type LightTrie struct {
-	trie *trie.SecureTrie
+func NewState(ctx context.Context, head *types.Header, odr OdrBackend) *state.StateDB {
+	state, _ := state.New(head.Root, NewStateDatabase(ctx, head, odr))
+	return state
+}
+
+func NewStateDatabase(ctx context.Context, head *types.Header, odr OdrBackend) state.Database {
+	return &odrDatabase{ctx, StateTrieID(head), odr}
+}
+
+type odrDatabase struct {
+	ctx     context.Context
+	id      *TrieID
+	backend OdrBackend
+}
+
+func (db *odrDatabase) OpenTrie(root common.Hash) (state.Trie, error) {
+	return &odrTrie{db: db, id: db.id}, nil
+}
+
+func (db *odrDatabase) OpenStorageTrie(addrHash, root common.Hash) (state.Trie, error) {
+	return &odrTrie{db: db, id: StorageTrieID(db.id, addrHash, root)}, nil
+}
+
+func (db *odrDatabase) CopyTrie(t state.Trie) state.Trie {
+	switch t := t.(type) {
+	case *odrTrie:
+		cpy := &odrTrie{db: t.db, id: t.id}
+		if t.trie != nil {
+			cpytrie := *t.trie
+			cpy.trie = &cpytrie
+		}
+		return cpy
+	default:
+		panic(fmt.Errorf("unknown trie type %T", t))
+	}
+}
+
+func (db *odrDatabase) ContractCode(addrHash, codeHash common.Hash) ([]byte, error) {
+	if codeHash == sha3_nil {
+		return nil, nil
+	}
+	if code, err := db.backend.Database().Get(codeHash[:]); err == nil {
```

Comment:
StateDB doesn't use cachingDB is the Database returned by state.NewDatabase. It's not used here.